### PR TITLE
Create classifier to differentiate case and control samples

### PR DESCRIPTION
**Pull request by ben-heil**
This PR has two main parts: a file of tensorflow models and the `classifier.py` program that trains them.
I will likely end up adding hyperparameters as arguments in a later PR. 

Is there a better way to import model architectures than using the Tensorflow/Keras Model API?

### CODE REVIEW

File: brdnet/classifier.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+import sys
+
+import numpy
+import pandas
+import tensorflow as tf
+
+import models
+
+# Set random seeds
+numpy.random.seed(42)
```
Comment (jjc2718) [inline brdnet/classifier.py]: Personally, I prefer to have the random seed be an argument with a default value, e.g. in your `argparse` code at the bottom of the script.

That way, if you want to quantify variability in your model later on, it's easy to run the script multiple times with different random seeds (and plot, for example, the mean/standard error in classification accuracy across seeds, rather than just a single value).

This can go in your future PR with the hyperparameters if you prefer, just want to make sure you don't miss it.

File: brdnet/classifier.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    -------
+    reduced_matrix: numpy.array
+        The result from translating expression_df into LV space
+    '''
+    expression_df = expression_df[expression_df.index.isin(Z_df.index)]
+
+    expression_df = expression_df.sort_index()
+
+    # Since the gene symbols are in alphabetical order and are identical between
+    # the two dataframes, we can drop the labels and create a numpy matrix to be multiplier by Z
```
Comment (jjc2718) [inline brdnet/classifier.py]: Typo multiplier -> multiplied (unless you're referring to Jackie's project here :smile: )

File: brdnet/classifier.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    expression_matrix = expression_df.values
+    Z_matrix = Z_df.values
+
+    reduced_matrix = numpy.matmul(expression_matrix.T, Z_matrix)
```
Comment (jjc2718) [inline brdnet/classifier.py:94]: You could try doing this directly in Pandas (using, e.g. [pandas.DataFrame.dot](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dot.html)). I think that should preserve the row/column labels, which might help with your TODO below.

I haven't used that function myself though, so you'll have to make sure it does what you want.

File: brdnet/classifier.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+    parser.add_argument('Z_file_path', help='Path to the PLIER matrix to be used to convert '
+                                            'the data into LV space')
+    parser.add_argument('healthy_file_path', help='Path to the tsv containing healthy gene '
+                                                  'expression')
+    parser.add_argument('disease_file_path', help='Path to the tsv containing unhealthy '
+                                                  'gene expression data')
+    parser.add_argument('--logdir', help='The directory to print tensorboard logs to',
+                        default='../logs')
+    parser.add_argument('--model', help='The name of the model to be used', default='MLP')
```
Comment (jjc2718) [inline brdnet/classifier.py]: Might be helpful to add a list of the models that are implemented here (e.g. using the `choices` argument in `argparse`). This isn't critical for this PR since your `validate_model_name` function will take care of erroneous model names, but might improve usability as you scale things up.

Off the top of my head, I don't know of a good way to automatically list, for example, all of the class names in your `models.py` file. Maybe one of the answers [here](https://stackoverflow.com/questions/17171878/how-to-find-the-list-of-all-the-class-name-in-a-file-in-python) would work?

Or you could just hardcode the model names, but then you have to update two places in the code every time you add a new model. Up to you.

**Review by jjc2718 — APPROVED** [review]
Looks good to me. Made some suggestions (and one typo fix), but all of this can wait until future PRs if you prefer.

> Is there a better way to import model architectures than using the Tensorflow/Keras Model API?

What do you mean by "import model architectures"? What you're doing seems fine to me, but maybe I don't understand your question.

File: brdnet/classifier.py (hunk 4 above)
Comment (ben-heil) [reply brdnet/classifier.py]: I pulled out the code from `write_invalid_model_error` that gets the model names to be its own function, which is now used to list the available models in both `write_invalid_model_error` and in the argparse help message

File: brdnet/classifier.py (hunk 3 above)
Comment (ben-heil) [reply brdnet/classifier.py:94]: That's a good idea, I'll definitely look into it once I start using the metadata
